File: src/services/jiraService.js
import axios from 'axios'

class JiraService {
  constructor() {
    this.baseURL = process.env.JIRA_BASE_URL || 'https://your-domain.atlassian.net'
    this.clientId = process.env.JIRA_CLIENT_ID
    this.clientSecret = process.env.JIRA_CLIENT_SECRET
    this.callbackURL = process.env.JIRA_CALLBACK_URL
    this.projectKey = process.env.JIRA_PROJECT_KEY || 'MVM'
    
    // Store access tokens (in production, use Redis or database)
    this.accessTokens = new Map()
    
    // OAuth endpoints
    this.authorizationURL = 'https://auth.atlassian.com/authorize'
    this.tokenURL = 'https://auth.atlassian.com/oauth/token'
    this.apiBaseURL = `${this.baseURL}/rest/api/3`
  }

  /**
   * Generate OAuth authorization URL
   * User should be redirected to this URL to start OAuth flow
   */
  getAuthorizationURL(state) {
    const params = new URLSearchParams({
      client_id: this.clientId,
      response_type: 'code',
      redirect_uri: this.callbackURL,
      scope: 'read:jira-work write:jira-work offline_access',
      state: state || this.generateRandomState()
    })
    return `${this.authorizationURL}?${params.toString()}`
  }

  /**
   * Exchange authorization code for access token
   */
  async exchangeCodeForToken(code) {
    try {
      const response = await axios.post(this.tokenURL, {
        grant_type: 'authorization_code',
        client_id: this.clientId,
        client_secret: this.clientSecret,
        code: code,
        redirect_uri: this.callbackURL
      }, {
        headers: {
          'Content-Type': 'application/json'
        }
      })

      const { access_token, refresh_token, expires_in } = response.data
      
      // Store token with expiration time
      const tokenData = {
        accessToken: access_token,
        refreshToken: refresh_token,
        expiresAt: Date.now() + (expires_in * 1000),
        expiresIn: expires_in
      }

      return tokenData
    } catch (error) {
      console.error('Error exchanging code for token:', error.response?.data || error.message)
      throw new Error(`Failed to exchange authorization code: ${error.message}`)
    }
  }

  /**
   * Refresh access token using refresh token
   */
  async refreshAccessToken(refreshToken) {
    try {
      const response = await axios.post(this.tokenURL, {
        grant_type: 'refresh_token',
        client_id: this.clientId,
        client_secret: this.clientSecret,
        refresh_token: refreshToken
      }, {
        headers: {
          'Content-Type': 'application/json'
        }
      })

      const { access_token, refresh_token, expires_in } = response.data
      
      return {
        accessToken: access_token,
        refreshToken: refresh_token || refreshToken, // Use new refresh token if provided
        expiresAt: Date.now() + (expires_in * 1000),
        expiresIn: expires_in
      }
    } catch (error) {
      console.error('Error refreshing access token:', error.response?.data || error.message)
      throw new Error(`Failed to refresh access token: ${error.message}`)
    }
  }

  /**
   * Store token for user session
   */
  setUserToken(userId, tokenData) {
    this.accessTokens.set(userId, tokenData)
  }

  /**
   * Get valid access token for user, refresh if needed
   */
  async getValidAccessToken(userId) {
    const tokenData = this.accessTokens.get(userId)
    
    if (!tokenData) {
      throw new Error('No token found for user')
    }

    // Check if token is expired or about to expire (within 5 minutes)
    if (Date.now() >= tokenData.expiresAt - 300000) {
      console.log('Access token expired or expiring soon, refreshing...')
      const newTokenData = await this.refreshAccessToken(tokenData.refreshToken)
      this.setUserToken(userId, newTokenData)
      return newTokenData.accessToken
    }

    return tokenData.accessToken
  }

  /**
   * Make authenticated API call to Jira
   */
  async apiCall(endpoint, method = 'GET', data = null, userId = null, accessToken = null) {
    try {
      // Get access token
      let token
      if (accessToken) {
        token = accessToken
      } else if (userId) {
        token = await this.getValidAccessToken(userId)
      } else {
        throw new Error('Either userId or accessToken must be provided')
      }

      const config = {
        method,
        url: `${this.apiBaseURL}${endpoint}`,
        headers: {
          'Authorization': `Bearer ${token}`,
          'Accept': 'application/json',
          'Content-Type': 'application/json'
        }
      }

      if (data && (method === 'POST' || method === 'PUT' || method === 'PATCH')) {
        config.data = data
      }

      const response = await axios(config)
      return response.data
    } catch (error) {
      console.error(`API call failed (${method} ${endpoint}):`, error.response?.data || error.message)
      throw error
    }
  }

  /**
   * Create a Jira issue
   */
  async createIssue(ticketData, userId, accessToken) {
    try {
      const issueData = {
        fields: {
          project: {
            key: this.projectKey
          },
          summary: ticketData.summary || 'Untitled Ticket',
          description: {
            type: 'doc',
            version: 3,
            content: [
              {
                type: 'paragraph',
                content: [
                  {
                    type: 'text',
                    text: ticketData.description || 'No description provided'
                  }
                ]
              }
            ]
          },
          issuetype: {
            name: 'Story'
          },
          priority: {
            name: ticketData.priority || 'Medium'
          }
        }
      }

      // Add assignee if provided
      if (ticketData.assignee && ticketData.assignee !== 'Unassigned') {
        issueData.fields.assignee = {
          name: ticketData.assignee
        }
      }

      // Add epic if provided
      if (ticketData.epic && ticketData.epic !== 'No Epic') {
        issueData.fields.customfield_10000 = {
          name: ticketData.epic
        }
      }

      // Add acceptance criteria as comment if provided
      if (ticketData.acceptanceCriteria && ticketData.acceptanceCriteria.length > 0) {
        issueData.fields.description.content.push({
          type: 'heading',
          attrs: { level: 2 },
          content: [
            {
              type: 'text',
              text: 'Acceptance Criteria'
            }
          ]
        })

        ticketData.acceptanceCriteria.forEach(criteria => {
          issueData.fields.description.content.push({
            type: 'bulletList',
            content: [
              {
                type: 'listItem',
                content: [
                  {
                    type: 'paragraph',
                    content: [
                      {
                        type: 'text',
                        text: criteria
                      }
                    ]
                  }
                ]
              }
            ]
          })
        })
      }

      const response = await this.apiCall('/issues', 'POST', issueData, userId, accessToken)
      return response
    } catch (error) {
      console.error('Error creating Jira issue:', error.message)
      throw new Error(`Failed to create Jira issue: ${error.message}`)
    }
  }

  /**
   * Create multiple Jira issues (batch operation)
   */
  async createMultipleIssues(tickets, userId, accessToken) {
    try {
      const results = []
      const errors = []

      for (const ticket of tickets) {
        try {
          const result = await this.createIssue(ticket, userId, accessToken)
          results.push({
            success: true,
            originalTicket: ticket,
            jiraTicket: result,
            issueKey: result.key,
            issueId: result.id
          })
        } catch (error) {
          errors.push({
            success: false,
            originalTicket: ticket,
            error: error.message
          })
        }
      }

      return {
        total: tickets.length,
        successful: results.length,
        failed: errors.length,
        results,
        errors
      }
    } catch (error) {
      console.error('Error in batch issue creation:', error.message)
      throw error
    }
  }

  /**
   * Update a Jira issue
   */
  async updateIssue(issueKey, updates, userId, accessToken) {
    try {
      const updateData = {
        fields: {}
      }

      if (updates.summary) {
        updateData.fields.summary = updates.summary
      }
      if (updates.description) {
        updateData.fields.description = {
          type: 'doc',
          version: 3,
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  type: 'text',
                  text: updates.description
                }
              ]
            }
          ]
        }
      }
      if (updates.priority) {
        updateData.fields.priority = { name: updates.priority }
      }
      if (updates.assignee) {
        updateData.fields.assignee = { name: updates.assignee }
      }

      const response = await this.apiCall(`/issues/${issueKey}`, 'PUT', updateData, userId, accessToken)
      return response
    } catch (error) {
      console.error(`Error updating Jira issue ${issueKey}:`, error.message)
      throw new Error(`Failed to update Jira issue: ${error.message}`)
    }
  }

  /**
   * Search for Jira issues using JQL
   */
  async searchIssues(jql, userId, accessToken) {
    try {
      const response = await this.apiCall(`/search?jql=${encodeURIComponent(jql)}`, 'GET', null, userId, accessToken)
      return response
    } catch (error) {
      console.error('Error searching Jira issues:', error.message)
      throw new Error(`Failed to search Jira issues: ${error.message}`)
    }
  }

  /**
   * Get Jira project details
   */
  async getProject(projectKey, userId, accessToken) {
    try {
      const response = await this.apiCall(`/projects/${projectKey}`, 'GET', null, userId, accessToken)
      return response
    } catch (error) {
      console.error(`Error getting project ${projectKey}:`, error.message)
      throw new Error(`Failed to get project details: ${error.message}`)
    }
  }

  /**
   * Get issue type schemes
   */
  async getIssueTypes(userId, accessToken) {
    try {
      const response = await this.apiCall('/issuetypes', 'GET', null, userId, accessToken)
      return response
    } catch (error) {
      console.error('Error getting issue types:', error.message)
      throw new Error(`Failed to get issue types: ${error.message}`)
    }
  }

  /**
   * Get issue by key
   */
  async getIssue(issueKey, userId, accessToken) {
    try {
      const response = await this.apiCall(`/issues/${issueKey}`, 'GET', null, userId, accessToken)
      return response
    } catch (error) {
      console.error(`Error getting issue ${issueKey}:`, error.message)
      throw new Error(`Failed to get issue: ${error.message}`)
    }
  }

  /**
   * Generate random state for OAuth
   */
  generateRandomState() {
    return Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15)
  }
}

export default new JiraService()

File: python-backend/services/jira_service.py
"""Jira OAuth2 integration service."""
from __future__ import annotations

import httpx
from typing import Any, Dict, List, Optional
from datetime import datetime, timedelta

from config.settings import settings


class JiraService:
    """Manages Jira OAuth2 authentication and ticket creation."""

    def __init__(self):
        """Initialize Jira service."""
        self.base_url = settings.JIRA_BASE_URL
        self.client_id = settings.JIRA_CLIENT_ID
        self.client_secret = settings.JIRA_CLIENT_SECRET
        self.callback_url = settings.JIRA_CALLBACK_URL
        self.auth_url = settings.JIRA_AUTH_URL
        self.token_url = settings.JIRA_TOKEN_URL
        
        # Session storage (in production, use Redis or database)
        self.sessions: Dict[str, Dict[str, Any]] = {}
        self.tokens: Dict[str, Dict[str, Any]] = {}

    async def get_auth_url(self) -> str:
        """Generate Jira OAuth2 authorization URL.

        Returns:
            Authorization URL for user redirect
        """
        params = {
            "client_id": self.client_id,
            "redirect_uri": self.callback_url,
            "response_type": "code",
            "scope": "read:jira-work write:jira-work manage:jira-project",
        }

        query_string = "&".join(f"{k}={v}" for k, v in params.items())
        auth_url = f"{self.auth_url}?{query_string}"

        return auth_url

    async def handle_callback(self, code: str, state: str) -> Dict[str, Any]:
        """Handle OAuth2 callback and exchange code for token.

        Args:
            code: Authorization code from Jira
            state: State parameter for CSRF protection

        Returns:
            Token data with expiration
        """
        if not code:
            raise ValueError("Authorization code missing")

        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    self.token_url,
                    data={
                        "grant_type": "authorization_code",
                        "client_id": self.client_id,
                        "client_secret": self.client_secret,
                        "code": code,
                        "redirect_uri": self.callback_url,
                    },
                    timeout=30,
                )

                if response.status_code != 200:
                    raise ValueError(f"OAuth2 token exchange failed: {response.text}")

                token_data = response.json()
                token_data["acquired_at"] = datetime.utcnow().isoformat()
                token_data["expires_at"] = (
                    datetime.utcnow() + timedelta(seconds=token_data.get("expires_in", 3600))
                ).isoformat()

                # Store token
                self.tokens[code] = token_data

                return token_data

        except httpx.RequestError as e:
            raise ValueError(f"Failed to exchange OAuth2 code: {str(e)}")

    async def create_tickets(self, tickets: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Create tickets in Jira.

        Args:
            tickets: List of ticket objects to create

        Returns:
            List of created ticket IDs and metadata
        """
        created_tickets = []

        for ticket in tickets:
            try:
                jira_ticket = self._convert_to_jira_format(ticket)
                # Note: Actual Jira API call would go here
                # For now, return mock response
                created_tickets.append(
                    {
                        "id": f"BA-{1001 + len(created_tickets)}",
                        "key": f"BA-{1001 + len(created_tickets)}",
                        "url": f"{self.base_url}/browse/BA-{1001 + len(created_tickets)}",
                        "originalId": ticket.get("id"),
                        "status": "created",
                    }
                )
            except Exception as e:
                created_tickets.append(
                    {
                        "originalId": ticket.get("id"),
                        "status": "failed",
                        "error": str(e),
                    }
                )

        return created_tickets

    async def status(self) -> Dict[str, Any]:
        """Get Jira connection status.

        Returns:
            Connection status and user info
        """
        if not self.tokens:
            return {
                "connected": False,
                "message": "No active Jira session",
            }

        # In production, verify token validity
        return {
            "connected": True,
            "message": "Connected to Jira",
            "sessions": len(self.tokens),
        }

    def _convert_to_jira_format(self, ticket: Dict[str, Any]) -> Dict[str, Any]:
        """Convert internal ticket format to Jira format.

        Args:
            ticket: Internal ticket object

        Returns:
            Jira-formatted ticket
        """
        return {
            "fields": {
                "project": {"key": "BA"},
                "summary": ticket.get("summary", "Untitled"),
                "description": ticket.get("description", ""),
                "issuetype": {"name": "Story"},
                "priority": self._map_priority(ticket.get("priority", "Medium")),
                "assignee": self._map_assignee(ticket.get("assignee")),
                "labels": self._extract_labels(ticket),
            }
        }

    @staticmethod
    def _map_priority(priority: str) -> Dict[str, str]:
        """Map internal priority to Jira priority.

        Args:
            priority: Internal priority level

        Returns:
            Jira priority mapping
        """
        priority_map = {
            "Critical": {"name": "Blocker"},
            "High": {"name": "High"},
            "Medium": {"name": "Medium"},
            "Low": {"name": "Low"},
        }
        return priority_map.get(priority, {"name": "Medium"})

    @staticmethod
    def _map_assignee(assignee: Optional[str]) -> Optional[Dict[str, str]]:
        """Map assignee to Jira format.

        Args:
            assignee: Assignee name or email

        Returns:
            Jira assignee mapping or None
        """
        if not assignee or assignee == "Unassigned":
            return None
        return {"name": assignee}

    @staticmethod
    def _extract_labels(ticket: Dict[str, Any]) -> List[str]:
        """Extract labels from ticket.

        Args:
            ticket: Ticket object

        Returns:
            List of labels
        """
        labels = []
        if epic := ticket.get("epic"):
            if epic != "No Epic":
                labels.append(f"epic-{epic}")

        if ticket_type := ticket.get("type"):
            labels.append(ticket_type.lower())

        return labels
